src/foam/dao/mongodb/MongoDAO.js

```diff
@@ -0,0 +1,348 @@
+        MongoCollection<BsonDocument> collection = getDatabase(x).getCollection(getCollectionName(), BsonDocument.class);
+        try {
+          collection.replaceOne(
+            buildIDFilter(x, obj),
+            toBson(x, obj),
+            UPSERT
+          );
+        } catch ( com.mongodb.MongoWriteException e ) {
+          getLogger().error("put_", obj.toSummary(), e.getMessage());
+          if ( e.getMessage().contains("duplicate key error") ) {
+            try {
+              long count = collection.countDocuments(buildIDFilter(x, obj));
+              getLogger().error("put_ duplicate count found", count);
+            } catch ( Exception ex ) {
+              getLogger().error("put_ duplicate count failed", ex.getMessage());
+            }
+          }
+          throw new RuntimeException(e);
+        } catch ( Exception e ) {
+          pm.error(x);
+          getLogger().error("put_", obj.toSummary(), e.getMessage(), e);
+          throw new RuntimeException(e);
+        }
+        pm.log(x);
+        return obj;
+      `
+    },
+    {
+      name: 'find_',
+      javaCode: `
+        PM pm = PM.create(x, getCollectionName(), "find_");
+        MongoCollection<BsonDocument> collection = getDatabase(x).getCollection(getCollectionName(), BsonDocument.class);
+        FObject result = null;
+        try {
+          var record = collection.find(buildIDFilter(x, id)).first();
+          if ( record != null ) {
+            result = parser_.get().parseString(record.toJson(), getOf().getObjClass());
+          }
+        } catch ( Exception e ) {
+          pm.error(x);
+          getLogger().error("find_", id, e.getMessage(), e);
+          throw new RuntimeException(e);
+        }
+        pm.log(x);
+        return result;
+      `
+    },
+    {
+      name: 'remove_',
+      javaCode: `
+        MongoCollection<BsonDocument> collection = getDatabase(x).getCollection(getCollectionName(), BsonDocument.class);
+        try {
+          collection.deleteOne(buildIDFilter(x, obj));
+        } catch ( Exception e ) {
+          getLogger().error("remove_", obj.toSummary(), e.getMessage());
+          throw new RuntimeException(e);
+        }
+        return obj;
+      `
+    },
+    {
+      name: 'select_',
+      javaCode: `
+        sink             = prepareSink(sink);
+        Sink decorated   = decorateSink_(sink, skip, limit, order, predicate);
+        Subscription sub = new Subscription();
+
+        MongoCollection<BsonDocument> collection = getDatabase(x).getCollection(getCollectionName(), BsonDocument.class);
+
+        try ( MongoCursor<BsonDocument> cursor = collection.find().cursor() ) {
+          while ( cursor.hasNext() ) {
+            if ( sub.getDetached() ) break;
+
+            FObject obj = parser_.get().parseString(cursor.next().toJson(), getOf().getObjClass());
+
+            if ( ( predicate == null ) || predicate.f(obj) ) {
+              decorated.put(obj, sub);
+            }
```

Comment:
The decorated sink should already be wrapped in PredicatedSink if there is a predicate.
```suggestion
            decorated.put(obj, sub);
```

Reply:
> The decorated sink should already be wrapped in PredicatedSink if there is a predicate.

This occurs in AbstractSink.decorateSink_ called on line 219.